src/shared/tpm2-util.c

```diff
@@ -457,26 +533,31 @@ int tpm2_context_new(const char *device, Tpm2Context **ret_context) {
         return 0;
 }
 
-static void tpm2_handle_flush(ESYS_CONTEXT *esys_context, ESYS_TR esys_handle) {
+static void tpm2_handle_cleanup(ESYS_CONTEXT *esys_context, ESYS_TR esys_handle, bool flush) {
+        TSS2_RC rc;
+
         if (!esys_context || esys_handle == ESYS_TR_NONE)
                 return;
 
-        TSS2_RC rc = sym_Esys_FlushContext(esys_context, esys_handle);
+        if (flush)
+                rc = sym_Esys_FlushContext(esys_context, esys_handle);
+        else
+                rc = sym_Esys_TR_Close(esys_context, &esys_handle);
         if (rc != TSS2_RC_SUCCESS) /* We ignore failures here (besides debug logging), since this is called
                                     * in error paths, where we cannot do anything about failures anymore. And
                                     * when it is called in successful codepaths by this time we already did
                                     * what we wanted to do, and got the results we wanted so there's no
                                     * reason to make this fail more loudly than necessary. */
-                log_debug("Failed to flush TPM handle, ignoring: %s", sym_Tss2_RC_Decode(rc));
+                log_debug("Failed to %s TPM handle, ignoring: %s", flush ? "flush" : "close", sym_Tss2_RC_Decode(rc));
 }
 
 Tpm2Handle *tpm2_handle_free(Tpm2Handle *handle) {
         if (!handle)
                 return NULL;
 
         _cleanup_(tpm2_context_unrefp) Tpm2Context *context = (Tpm2Context*)handle->tpm2_context;
-        if (context && !handle->keep)
-                tpm2_handle_flush(context->esys_context, handle->esys_handle);
+        if (context)
+                tpm2_handle_cleanup(context->esys_context, handle->esys_handle, !handle->no_flush);
```

Comment:
so, sometimes we work with an inverted "no_flush" bool and sometimes with a regular "flush" bool. That's confusing. Please stick to non-inverted bools, out of principle, except if there's a strong reason.

Reply:
ack, updated the field to 'flush'